### PR TITLE
Add Rpaas_ResourceProvisioningState, improve TopLevelResourcesListByResourceGroup 

### PR DESCRIPTION
**Pull request by jianyexi**
doc:
https://github.com/Azure/azure-rest-api-specs/pull/12268

### CODE REVIEW

File: src/typescript/azure-openapi-validator/rules/Rpaas_ResourceProvisioningState.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  appliesTo_JsonQuery: "$",
+  *run(doc, node, path) {
+    const msg = `The resource {0} is defined without 'provisioningState' in properties bag, consider adding the provisioningState for it.`
+    const utils = new ResourceUtils(doc)
+    const allResources = utils.getAllResource()
+    for (const resource of allResources) {
+      const model = utils.getResourceByName(resource)
+      const properties = utils.getPropertyOfModel(model,"properties")
+      let hasProvisioningState = false
+      if (properties && (!properties.type || properties.type === "object")) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/Rpaas_ResourceProvisioningState.ts:26]: If type undefined, is it valid?

Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/Rpaas_ResourceProvisioningState.ts:26]: yes,  undefined type  equals to  "object"

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 2]
```diff
@@ -195,6 +199,42 @@ export class ResourceUtils {
     }
     return topLevelModels
   }
+
+  public getTopLevelResourcesByRG() {
+    const fullModels = this.getOperationGetResponseResources()
+    const topLevelModels = new Set<string>()
+    for (const entry of fullModels.entries()) {
+      const paths = entry[1]
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:207]: Could you pls add comments here? Why '1'? looks like a magic number

Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:207]: The entries() return  k-v pair array which contains 2 items , not magic number

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 3]
```diff
@@ -455,20 +498,29 @@ export class ResourceUtils {
     }
     let model = sourceModel
     if (sourceModel.$ref) {
-      model = this.getResourceByName(this.stripDefinitionPath(sourceModel))
+      model = this.getUnwrappedModel(sourceModel.$ref)
+    }
+    if (!model) {
+      return undefined
     }
     if (model.properties) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:506]: model.properties?.propertyName

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts [hunk 4]
```diff
@@ -455,20 +498,29 @@ export class ResourceUtils {
...
-          return ref ? this.dereference(ref) : property
+        if (element.$ref) {
+          const property = this.getPropertyOfModelName(this.stripDefinitionPath(element.$ref), propertyName)
+          if (property) {
+            return this.getUnwrappedModel(property)
+          }
+        }
+        else {
+          const property = this.getPropertyOfModel(element,propertyName)
+          if (property) {
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts:521]: Why not directly return property?

Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts]: good idea, fixed

File: src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts (hunk 3 above)
Comment (jianyexi) [reply src/typescript/azure-openapi-validator/rules/utilities/resourceUtils.ts]: No, 'propertyName' is variable

File: src/typescript/azure-openapi-validator/rules/Rpaas_ResourceProvisioningState.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import { MergeStates, OpenApiTypes, rules } from "../rule"
+import { ResourceUtils } from "./utilities/resourceUtils"
+import { JsonPath } from "../../jsonrpc/types"
+export const Rpaas_ResourceProvisioningState: string = "Rpaas_ResourceProvisioningState"
+
+rules.push({
```
Comment (ruowan) [inline src/typescript/azure-openapi-validator/rules/Rpaas_ResourceProvisioningState.ts:10]: Is there any doc about why the rule is reasonable?

**Review by ruowan — APPROVED** [review]

**Comment by jianyexi** [comment]
@sanjaiganesh could you help review the PR ?

**Comment by jianyexi** [comment]
> @sanjaiganesh could you help review the PR ?

Run regression test against rpaasMaster branch , there are totally 17 Rpaas_ResourceProvisioningState rule errors 
see attached file
[linting_all_rpaas_readmes.test.log](https://github.com/Azure/azure-openapi-validator/files/5785475/linting_all_rpaas_readmes.test.log)

**Review by ruowan — APPROVED** [review]

**Pull request merged**